File: python/977_SquaresOfASortedArray.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        start, end, i = 0, len(nums) - 1, len(nums) - 1
        res = [0 for _ in range(len(nums))]
        while start <= end:
            if abs(nums[start]) > abs(nums[end]):
                res[i] = nums[start] * nums[start]
                start += 1
            else:
                res[i] = nums[end] * nums[end]
                end -= 1
            i -= 1
        return res

class Solution:
    def sortedSquares(self, A: List[int]) -> List[int]:
        idx, neg_squares = 0, []
        while idx < len(A) and A[idx] < 0:
            neg_squares.append(A[idx]**2)
            idx += 1

        squares = [n**2 for n in A[idx:]]
        
        i = 0
        while i < len(squares):
            if neg_squares and squares[i] >= neg_squares[-1]:
                squares.insert(i, neg_squares[-1])
                del neg_squares[-1]
            i += 1
        
        if not squares:
            return neg_squares[::-1]
        elif neg_squares and neg_squares[-1] > squares[-1]: 
            while neg_squares:
                squares.append(neg_squares[-1])
                del neg_squares[-1]
        
        return squares
        
        
        

File: cpp/977_SquaresOfASortedArray.cpp
class Solution {
public:
    vector<int> sortedSquares(vector<int>& nums) {
        vector<int> res(nums.size());
        int start = 0, end = nums.size() - 1, i = end;
        while (start <= end) {
            res[i--] = abs(nums[start]) > abs(nums[end]) ? nums[start] * nums[start++] : nums[end] * nums[end--];
        }
        return res;
    }
};

class Solution {
public:
    vector<int> sortedSquares(vector<int>& A) {
        vector<int> neg_stack;
        int idx = 0; 
        int num_negs = 0;
        
        while (idx < A.size() && A[idx] < 0) {
            neg_stack.push_back(A[idx]*A[idx]);
            ++idx, ++num_negs;
        }

        vector<int> res;
        for (int i = num_negs; i < A.size(); ++i) {
            int square = A[i] * A[i];
            while (!neg_stack.empty() && square > neg_stack.back()) {
                res.push_back(neg_stack.back());
                neg_stack.pop_back();
            }
            res.push_back(square);
        }
        while (!neg_stack.empty()) {
            res.push_back(neg_stack.back());
            neg_stack.pop_back();
        }
        return res;
    }
};
